lib/Sema/CSSimplify.cpp

```diff
@@ -3449,24 +3448,104 @@ ConstraintSystem::simplifyMemberConstraint(ConstraintKind kind,
     instanceTy = MTT->getInstanceType();
   
   // Value member lookup has some hacks too.
-  if (shouldAttemptFixes() && baseObjTy->getOptionalObjectType()) {
```

Comment:
I'd really appreciate if the whole original block of code was removed from the `simplifyMemberConstraint` and new logic would be somehow integrated into `CSDiag`, I think it's better place to do this kind of work since we'd be able to evaluate optional chaining per overload choice without risking blowing out the stack and we'd know exactly where in the expression chain it is. Currently we have some logic to handle optional unwraps here and in `CSDiag` it would be great to unify them.

I'd be great if you could overload `takeNon` and `takeOpt` since that makes things more interesting.